database/src/service/write.rs

```diff
@@ -48,22 +50,54 @@ pub struct DatabaseWriteHandle {
     /// Sender channel to the database write thread-pool.
     ///
     /// We provide the response channel for the thread-pool.
-    pub(super) sender: crossbeam::channel::Sender<(WriteRequest, ResponseSend)>,
+    pub(super) sender: crossbeam::channel::Sender<(WriteRequest, ResponseSender)>,
+}
+
+impl DatabaseWriteHandle {
+    /// Initialize the single `DatabaseWriter` thread.
+    #[cold]
+    #[inline(never)] // Only called once.
+    pub(super) fn init(db: Arc<ConcreteEnv>) -> Self {
+        // Initialize `Request/Response` channels.
+        let (sender, receiver) = crossbeam::channel::unbounded();
+
+        // Spawn the writer.
+        std::thread::Builder::new()
+            .name(WRITER_THREAD_NAME.into())
+            .spawn(move || {
+                let this = DatabaseWriter { receiver, db };
+                DatabaseWriter::main(this);
+            })
+            .unwrap();
+
+        Self { sender }
+    }
 }
 
 impl tower::Service<WriteRequest> for DatabaseWriteHandle {
     type Response = Response;
     type Error = RuntimeError;
-    type Future = ResponseRecv;
+    type Future = ResponseReceiver;
 
     #[inline]
-    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        todo!()
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // If there are previous write requests not fulfilled, wait.
+        if self.sender.is_empty() {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Pending
+        }
```

Comment:
Not sure if this is needed.

Reply:
The channel itself acts as a backlog since it's unbounded, so this could be removed?

Reply:
Yeah I don't think this is needed for the write service.

Plus when you return `Poll::Pending` from a future/ function returning `Poll` you MUST ensure the `Context` is woken up when the future becomes ready, here you drop the context so when the channle is full any write requests would just wait forever even if the channel is no longer empty